okhttp/src/main/java/okhttp3/internal/ws/WebSocketExtensions.kt

```diff
@@ -0,0 +1,172 @@
+          continue // Not a header we're interested in.
+        }
+        val header = response.headers.value(i)
+
+        // Parse each extension.
+        var pos = 0
+        while (pos < header.length) {
+          val extensionEnd = header.delimiterOffset(',', pos)
+          val extensionTokenEnd = header.delimiterOffset(';', pos, extensionEnd)
+          val extensionToken = header.trimSubstring(pos, extensionTokenEnd)
+          pos = extensionTokenEnd + 1
+
+          when {
+            extensionToken.equals("permessage-deflate", ignoreCase = true) -> {
+              if (compressionEnabled) unexpectedValues = true // Repeated extension!
+              compressionEnabled = true
+
+              // Parse each permessage-deflate parameter.
+              while (pos < extensionEnd) {
+                val parameterEnd = header.delimiterOffset(';', pos, extensionEnd)
+                val equals = header.delimiterOffset('=', pos, parameterEnd)
+                val name = header.trimSubstring(pos, equals)
+                val value = if (equals < parameterEnd) {
+                  header.trimSubstring(equals + 1, parameterEnd).removeSurrounding("\"")
+                } else {
+                  null
+                }
+                pos = parameterEnd + 1
+                when {
+                  name.equals("client_max_window_bits", ignoreCase = true) -> {
+                    if (clientMaxWindowBits != null) unexpectedValues = true // Repeated parameter!
+                    clientMaxWindowBits = value?.toIntOrNull()
+                    if (clientMaxWindowBits == null) unexpectedValues = true // Not an int!
+                  }
+                  name.equals("client_no_context_takeover", ignoreCase = true) -> {
+                    if (clientNoContextTakeover) unexpectedValues = true // Repeated parameter!
+                    if (value != null) unexpectedValues = true // Unexpected value!
+                    clientNoContextTakeover = true
+                  }
+                  name.equals("server_max_window_bits", ignoreCase = true) -> {
+                    if (serverMaxWindowBits != null) unexpectedValues = true // Repeated parameter!
+                    serverMaxWindowBits = value?.toIntOrNull()
+                    if (serverMaxWindowBits == null) unexpectedValues = true // Not an int!
+                  }
+                  name.equals("server_no_context_takeover", ignoreCase = true) -> {
+                    if (serverNoContextTakeover) unexpectedValues = true // Repeated parameter!
+                    if (value != null) unexpectedValues = true // Unexpected value!
+                    serverNoContextTakeover = true
+                  }
+                  else -> {
+                    unexpectedValues = true // Unexpected parameter.
+                  }
+                }
+              }
+            }
+
+            else -> {
+              unexpectedValues = true // Unexpected extension.
```

Comment:
WDYT about making `unexpectedValues` a data class, and save the message that is currently in comments to it? Along with the content of `Sec-WebSocket-Extensions` header that could not be parsed. 

This way we could throw a meaningful exception to the user.

Reply:
I don't think we're going to need it in practice. If we want the user to understand the problem, we should just include the unparsed header.